bfvmm/src/hve/arch/intel_x64/ept/memory_map.cpp

```diff
@@ -0,0 +1,372 @@
+}
+
+epte_t &
+memory_map::map_pdpte_to_page(gpa_t gpa, hpa_t hpa)
+{
+    auto &pml4e = this->gpa_to_pml4e(gpa);
+    if (!epte::is_present(pml4e)) {
+        auto pdpt_hpa = this->allocate_page_table();
+
+        epte::read_access::enable(pml4e);
+        epte::write_access::enable(pml4e);
+        epte::set_hpa(pml4e, pdpt_hpa);
+    }
+
+    auto &pdpte = this->gpa_to_pdpte(gpa, pml4e);
+    if (epte::is_present(pdpte)) {
+        throw std::runtime_error("map_pdpte_to_page: failed to map gpa, gpa is "
+            "already mapped at the 1GB level");
+    }
+
+    this->map_entry_to_page_frame(pdpte, hpa);
+    return pdpte;
+}
+
+epte_t &
+memory_map::map_pde_to_page(gpa_t gpa, hpa_t hpa)
+{
+    auto &pml4e = this->gpa_to_pml4e(gpa);
+    if (!epte::is_present(pml4e)) {
+        auto pdpt_hpa = this->allocate_page_table();
+
+        epte::read_access::enable(pml4e);
+        epte::write_access::enable(pml4e);
+        epte::set_hpa(pml4e, pdpt_hpa);
+    }
+
+    auto &pdpte = this->gpa_to_pdpte(gpa, pml4e);
+    if (epte::entry_type::is_enabled(pdpte)) {
+        throw std::runtime_error("map_pde_to_page: failed to map gpa, gpa is "
+            "already mapped at the 1GB level");
+    }
+    if (!epte::is_present(pdpte)) {
+        this->allocate_page_table(pdpte);
+    }
+
+    auto &pde = this->gpa_to_pde(gpa, pdpte);
+    if (epte::is_present(pde)) {
+        throw std::runtime_error("map_pde_to_page: failed to map gpa, gpa is "
+            "already mapped at the 2MB level");
+    }
+
+    this->map_entry_to_page_frame(pde, hpa);
+    return pde;
+}
+
+epte_t &
+memory_map::map_pte_to_page(gpa_t gpa, hpa_t hpa)
+{
+    auto &pml4e = this->gpa_to_pml4e(gpa);
+    if (!epte::is_present(pml4e)) {
+        auto pdpt_hpa = this->allocate_page_table();
+
+        epte::read_access::enable(pml4e);
+        epte::write_access::enable(pml4e);
+        epte::set_hpa(pml4e, pdpt_hpa);
+    }
+
+    auto &pdpte = this->gpa_to_pdpte(gpa, pml4e);
+    if (epte::entry_type::is_enabled(pdpte)) {
+        throw std::runtime_error("map_pte_to_page: failed to map gpa, gpa is "
+            "already mapped at the 1GB level");
+    }
+    if (!epte::is_present(pdpte)) {
+        this->allocate_page_table(pdpte);
+    }
+
+    auto &pde = this->gpa_to_pde(gpa, pdpte);
+    if (epte::entry_type::is_enabled(pde)) {
+        throw std::runtime_error("map_pte_to_page: failed to map gpa, gpa is "
+            "already mapped at the 2MB level");
+    }
+    if (!epte::is_present(pde)) {
+        this->allocate_page_table(pde);
+    }
+
+    auto &pte = this->gpa_to_pte(gpa, pde);
+    if (epte::is_present(pte)) {
+        throw std::runtime_error("map_pte_to_page: failed to map gpa, gpa is "
+            "already mapped at the 4KB level");
+    }
+
+    this->map_entry_to_page_frame(pte, hpa);
+    return pte;
+}
+
+std::vector<memory_descriptor>
+memory_map::to_mdl(std::vector<memory_descriptor> mdl, epte_t *page_table) const
```

Comment:
rework this. You should never pass a class via copy. Instead pass by reference. In this case, since you are constructing a vector, pass by non-const, so that you do not need to copy on line 364. 

In general.... when you are working with a class (especially a data structure), be VERY careful and mindful of the different between a copy, move, etc... 

Give this a watch as it is really helpful:
https://channel9.msdn.com/Shows/Going+Deep/Cpp-and-Beyond-2012-Scott-Meyers-Universal-References-in-Cpp11